In [1]:
import pandas as pd

In [2]:
# save filepath to variable for easier access
books_file_path = '../data/BX-Books.csv'
# read the data and store data in DataFrame titled books_data
books_data = pd.read_csv((books_file_path), sep=';', error_bad_lines=False, encoding='ISO-8859-1')
#.decode('ISO-8859-1') 
# print a summary of the data
books_data.describe()

/usr/local/Caskroom/miniforge/base/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkippin

ISBN       BookTitle      Book-Author  Year-Of-Publication  \
count       271360          271360           271359               271360   
unique      271360          242135           102023                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

In [49]:
# save filepath to variable for easier access
bookratings_file_path = '../data/BX-Book-Ratings.csv'
# read the data and store data in DataFrame titled bookratings_data
bookratings_data = pd.read_csv((bookratings_file_path), sep=';', error_bad_lines=False, encoding='ISO-8859-1')
#.decode('ISO-8859-1') 
# print a summary of the data
bookratings_data.columns

/usr/local/Caskroom/miniforge/base/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


Index(['UserID', 'ISBN', 'BookRating'], dtype='object')

In [4]:
# save filepath to variable for easier access
users_file_path = '../data/BX-Users.csv'
# read the data and store data in DataFrame titled users_data
users_data = pd.read_csv((users_file_path), sep=';', error_bad_lines=False, encoding='ISO-8859-1')
#.decode('ISO-8859-1') 
# print a summary of the data
users_data.describe()

/usr/local/Caskroom/miniforge/base/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


UserID            Age
count  278858.00000  168096.000000
mean   139429.50000      34.751434
std     80499.51502      14.428097
min         1.00000       0.000000
25%     69715.25000      24.000000
50%    139429.50000      32.000000
75%    209143.75000      44.000000
max    278858.00000     244.000000

In [152]:
# print(books_data['BookTitle'].where(books_data['BookTitle'] == 'Cat &amp; Mouse (Alex Cross Novels)'))

title = books_data.loc[books_data['BookTitle'].str.contains("Seabiscuit")]

title

ISBN                       BookTitle        Book-Author  \
38      0449005615  Seabiscuit: An American Legend  LAURA HILLENBRAND   
74      0345465083                      Seabiscuit  LAURA HILLENBRAND   
12067   0375502912  Seabiscuit: An American Legend  LAURA HILLENBRAND   
248855  0760754306             Come on Seabiscuit!        Ralph Moody   
270766  0345471156      Seabiscuit: The Screenplay          GARY ROSS   

       Year-Of-Publication           Publisher  \
38                    2002    Ballantine Books   
74                    2003    Ballantine Books   
12067                 2001        Random House   
248855                2004  Barnes &amp; Noble   
270766                2003    Ballantine Books   

                                              Image-URL-S  \
38      http://images.amazon.com/images/P/0449005615.0...   
74      http://images.amazon.com/images/P/0345465083.0...   
12067   http://images.amazon.com/images/P/0375502912.0...   
248855  http://images.amazon.com/images/P/0760754306.0...   
270766  http://images.amazon.com/images/P/0345471156.0...   

                                              Image-URL-M  \
38      http://images.amazon.com/images/P/0449005615.0...   
74      http://images.amazon.com/images/P/0345465083.0...   
12067   http://images.amazon.com/images/P/0375502912.0...   
248855  http://images.amazon.com/images/P/0760754306.0...   
270766  http://images.amazon.com/images/P/0345471156.0...   

                                              Image-URL-L  
38      http://images.amazon.com/images/P/0449005615.0...  
74      http://images.amazon.com/images/P/0345465083.0...  
12067   http://images.amazon.com/images/P/0375502912.0...  
248855  http://images.amazon.com/images/P/0760754306.0...  
270766  http://images.amazon.com/images/P/0345471156.0...

In [153]:
bookratings_data

title = bookratings_data.loc[bookratings_data['ISBN'] == '0449005615']
title

UserID        ISBN  BookRating
2813     277722  0449005615          10
2836     277744  0449005615           0
9608         26  0449005615           9
10423       367  0449005615           9
11638       843  0449005615           0
...         ...         ...         ...
1117543  268110  0449005615           0
1121537  269280  0449005615           9
1131818  271448  0449005615           0
1133114  271720  0449005615           7
1142302  274363  0449005615           8

[246 rows x 3 columns]

In [7]:
users_data.columns

Index(['UserID', 'Location', 'Age'], dtype='object')

In [9]:
users_data.sort_values('UserID', axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last', ignore_index=False, key=None)

UserID                            Location   Age
0            1                  nyc, new york, usa   NaN
1            2           stockton, california, usa  18.0
2            3     moscow, yukon territory, russia   NaN
3            4           porto, v.n.gaia, portugal  17.0
4            5  farnborough, hants, united kingdom   NaN
...        ...                                 ...   ...
278853  278854               portland, oregon, usa   NaN
278854  278855  tacoma, washington, united kingdom  50.0
278855  278856           brampton, ontario, canada   NaN
278856  278857           knoxville, tennessee, usa   NaN
278857  278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [10]:
bookratings_data.sort_values('UserID', axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last', ignore_index=False, key=None)

UserID        ISBN  BookRating
9561       2  0195153448           0
9562       7   034542252           0
9572       8  0771025661           0
9580       8  1881320189           7
9579       8  1575663937           6
...      ...         ...         ...
9560  278854  0553579606           8
9554  278854  0375703063           7
9553  278854  0316184152           7
9555  278854  042516098X           7
9559  278854  0553578596           0

[1149780 rows x 3 columns]

In [ ]:
bookratings_data.sort_values('')

In [20]:
book_rating = pd.merge(bookratings_data, books_data, on='ISBN')
cols = ['Year-Of-Publication', 'Publisher', 'Book-Author', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L']
book_rating.drop(cols, axis=1, inplace=True)

rating_count = (book_rating.
    groupby(by = ['BookTitle'])['BookRating'].
    count().
    reset_index().
    rename(columns = {'BookRating': 'RatingCount_book'})
    [['BookTitle', 'RatingCount_book']]
    )

rating_count.head()

BookTitle  RatingCount_book
0   A Light in the Storm: The Civil War Diary of ...                 4
1                              Always Have Popsicles                 1
2               Apple Magic (The Collector's series)                 1
3   Ask Lily (Young Women of Faith: Lily Series, ...                 1
4   Beyond IBM: Leadership Marketing and Finance ...                 1

In [22]:
threshold = 25
rating_count = rating_count.query('RatingCount_book >= @threshold')
rating_count.shape

(5854, 2)

In [23]:
rating_count

BookTitle  RatingCount_book
75                                           'Salem's Lot                47
203                                        10 Lb. Penalty                61
422                                        101 Dalmatians                37
673                       14,000 Things to Be Happy About                28
697                                    16 Lighthouse Road                65
...                                                   ...               ...
240860                                \O\" Is for Outlaw"               269
240874  \Surely You're Joking, Mr. Feynman!\": Adventu...                57
240955                                                  e                44
240978                            iI Paradiso Degli Orchi                27
241008                                           stardust                51

[5854 rows x 2 columns]

In [47]:
threshold = 25
mthreshold = 200
rating_count = rating_count.query('RatingCount_book >= @threshold and RatingCount_book <= @mthreshold')
rating_count.shape

(5561, 2)

In [48]:
rating_count.sort_values('RatingCount_book', ascending=False).head(60)

Book-Title  RatingCount_book
85713   Hot Six : A Stephanie Plum Novel (A Stephanie ...               200
175510                              The Angel of Darkness               200
151644                                        Rose Madder               200
159775  Shopaholic Takes Manhattan (Summer Display Opp...               200
88623                     I Know Why the Caged Bird Sings               199
41608   Cruel &amp; Unusual (Kay Scarpetta Mysteries (...               199
156021                                 Sea Glass: A Novel               199
175222      The Amazing Adventures of Kavalier &amp; Clay               198
32206                 Cat &amp; Mouse (Alex Cross Novels)               198
178608                        The Bonfire of the Vanities               197
207582                                   The Simple Truth               197
91010                              In Her Shoes : A Novel               197
127366                                              Naked               196
150421                                            Riptide               196
184212                          The Deep End of the Ocean               196
238634                                  Wuthering Heights               195
19594                                        Bag of Bones               195
183208                    The Crimson Petal and the White               195
37118                                       Cold Mountain               194
236087                                    Winter Solstice               194
167300                                       Still Waters               194
198265                                The Mists of Avalon               194
193259                                         The Jester               194
30547                                    Call of the Wild               193
93012                                           Intensity               193
64056                                           Faking It               193
209676  The Tale of the Body Thief (Vampire Chronicles...               193
188304                                    The Fourth Hand               192
209478                                         The Switch               192
4450    A Lesson Before Dying (Vintage Contemporaries ...               192
34926                                         Chosen Prey               192
7743                                    A Wrinkle in Time               192
143455                                      Prey: A Novel               191
21876                                             Beloved               191
66361                                        Final Target               191
61064                                                Envy               190
24959                                           Bleachers               189
64167                                      Falling Angels               189
143451                                               Prey               188
124206                                         Mr. Murder               188
180635        The Celestine Prophecy (Celestine Prophecy)               188
55791                   Drowning Ruth (Oprah's Book Club)               187
11141                                 All Around the Town               187
156101                                         Seabiscuit               187
217516                                            Thinner               186
104004                                  Last Man Standing               186
112320  Love in the Time of Cholera (Penguin Great Boo...               186
145615                                           Quentins               185
225076                               Under the Tuscan Sun               184
213804                                         The Winner               184
112650                        Loves Music, Loves to Dance               184
11369   All That Remains (Kay Scarpetta Mysteries (Pap...               184
190470             

# Non Fiction

## Autobiography
88623	I Know Why the Caged Bird Sings	199

# Humor and Satire
85713	Hot Six : A Stephanie Plum Novel (A Stephanie ...	200


# Fiction


## Thrillers and Suspense
175510	The Angel of Darkness	200
32206	Cat &amp; Mouse (Alex Cross Novels)	198
193259	The Jester	194


### Crime Fiction
41608	Cruel &amp; Unusual (Kay Scarpetta Mysteries (...	199
184212	The Deep End of the Ocean	196
167300	Still Waters	194

### Action and Adventure
207582	The Simple Truth	197
150421	Riptide	196
30547	Call of the Wild	193


## Horror and Fantasy
151644	Rose Madder	200
19594	Bag of Bones	195
93012	Intensity	193
209676	The Tale of the Body Thief (Vampire Chronicles...	193
143455	Prey: A Novel	191


## Women's Fiction (Humor)
159775	Shopaholic Takes Manhattan (Summer Display Opp...	200


## Women's Fiction / Contemporary
91010	In Her Shoes : A Novel	197
236087	Winter Solstice	194


## Drama / Contemporary
188304	The Fourth Hand	192


## Historical Fiction

### Humor
175222	The Amazing Adventures of Kavalier &amp; Clay	198

### Romance
183208	The Crimson Petal and the White	195
37118	Cold Mountain	194
156021	Sea Glass: A Novel	199


### Fantasy
198265	The Mists of Avalon	194


## Satire / Humor
178608	The Bonfire of the Vanities	197


## Essay Collection / Humor
127366	Naked	196


## Romance
209478	The Switch	192




34926	Chosen Prey	192
7743	A Wrinkle in Time	192

21876	Beloved	191
66361	Final Target	191
61064	Envy	190
24959	Bleachers	189
64167	Falling Angels	189
143451	Prey	188
124206	Mr. Murder	188
180635	The Celestine Prophecy (Celestine Prophecy)	188
55791	Drowning Ruth (Oprah's Book Club)	187
11141	All Around the Town	187
156101	Seabiscuit	187
217516	Thinner	186
104004	Last Man Standing	186
112320	Love in the Time of Cholera (Penguin Great Boo...	186
145615	Quentins	185
225076	Under the Tuscan Sun	184
213804	The Winner	184
112650	Loves Music, Loves to Dance	184
11369	All That Remains (Kay Scarpetta Mysteries (Pap...	184
190470	The Guardian	183
39675	Coraline	183
120184	Midnight Bayou	183
23720	Big Stone Gap: A Novel (Ballantine Reader's Ci...	182
95102	Jack &amp; Jill (Alex Cross Novels)	182
143496	Pride and Prejudice	182
222259	Tribulation Force: The Continuing Drama of Tho...	182
162395	Smilla's Sense of Snow	181

In [49]:
import sqlalchemy